src/clients/HttpTunneler.cc

```diff
@@ -122,23 +123,66 @@ Http::Tunneler::startReadingResponse()
 }
 
 void
-Http::Tunneler::writeRequest()
+Http::Tunneler::buildHttpHeaders(HttpHeader &hdr_out)
 {
-    debugs(83, 5, connection);
-
-    HttpHeader hdr_out(hoRequest);
     Http::StateFlags flags;
     flags.peering = true;
     // flags.tunneling = false; // the CONNECT request itself is not tunneled
     // flags.toOrigin = false; // the next HTTP hop is a non-originserver peer
+    if (request->method == Http::METHOD_CONNECT) {
+        // We are forwarding an HTTP CONNECT request
+        HttpStateData::httpBuildRequestHeader(request.getRaw(),
+                                              nullptr, // StoreEntry
+                                              al,
+                                              &hdr_out,
+                                              flags);
+    } else {
+        // We need to create a new CONNECT request to forward the HttpRequest object
+
+        // The new HttpRequest object shares the same features with the
+        // origin, including the masterXaction object.
+        // XXX: Should the peerRequest be handled as a new HTTP transaction?
+        HttpRequest::Pointer peerRequest = new HttpRequest(request->masterXaction);
+        peerRequest->initHTTP(Http::METHOD_CONNECT, AnyP::PROTO_HTTP, "http", url.c_str());
+
+        // OK about auth and client related info, but what about
```

Comment:
Why are they OK? Is that because HttpStateData::httpBuildRequestHeader() understands the difference between client-Squid authentication and Squid-peer authentication, handling things like credentials forwarding correctly already?

Reply:
Yes. Moreover the authentication info copied from CONNECT request to bumped GET requests.

Reply:
> copied from CONNECT request to bumped GET requests

There is no such copying _here_ though, right? We just created a CONNECT request (i.e., `peerRequest`) and that peerRequest object will be destroyed at the end of this function. You are talking about some other copying that happens somewhere else, for a different CONNECT request, right?

Reply:
The HttpRequest::inheritProperties copies the HttpRequest::auth_user_request between others.
If I am not wrong these are the same with the original CONNECT request (which bumped). 

I think this is correct. Imagine the original CONNECT request it may include cretendials need to be forwarded to the upstream proxy.  And if my patch does not forward them correctly we should fix it to forward them.

Reply:
> original CONNECT

The code in question is prefixed with `We need to create a new CONNECT request` and, essentially, the negation of `We are forwarding an HTTP CONNECT request`. Thus, AFAICT, there is no "original CONNECT" here. There is only "GET https". Please correct me if I am wrong.

Reply:
>Thus, AFAICT, there is no "original CONNECT" here. 

Yes there is not "original CONNECT". I was refering to the initial client-to-squid CONNECT request which bumped. But we do not have any access to this CONNECT request any more.

> There is only "GET https". Please correct me if I am wrong.

You are right.

Reply:
> there is not "original CONNECT". I was refering to the initial client-to-squid CONNECT request which bumped.

To me, "the initial client-to-squid CONNECT request which bumped" _is_ the "original CONNECT". 

1. The code comments make me think that the `else` clause deals exclusively with "GET https" (no CONNECT at all).
1. Your responses make me think that the `else` clause deals with requests inside a step1-bumped CONNECT tunnel.

What does the `else` clause actually deal with?

Reply:
The "else" clause intended to deal with both cases. However currently deals only with the step1-bumped CONNECT tunnels. Currently ee are not using Http::Tunneler for "get https://" client requests.